src/assign.c

```diff
@@ -942,37 +942,51 @@ void memrecycle(SEXP target, SEXP where, int start, int len, SEXP source)
   UNPROTECT(protecti);
 }
 
-SEXP allocNAVector(SEXPTYPE type, R_len_t n)
+void writeNA(SEXP v, const int from, const int n)
 {
-  // an allocVector following with initialization to NA since a subassign to a new column using :=
-  // routinely leaves untouched items (rather than 0 or "" as allocVector does with its memset)
-  // We guess that author of allocVector would have liked to initialize with NA but was prevented since memset
-  // is restricted to one byte.
-  SEXP v = PROTECT(allocVector(type, n));
-  switch(type) {
+  const int to = from-1+n;  // together with <=to below with writing NA to position 2147483647 in mind
+  switch(TYPEOF(v)) {
   case LGLSXP : {
     Rboolean *vd = (Rboolean *)LOGICAL(v);
-    for (int i=0; i<n; i++) vd[i] = NA_LOGICAL;
+    for (int i=from; i<=to; ++i) vd[i] = NA_LOGICAL;
   } break;
   case INTSXP : {
+    // same whether factor or not
     int *vd = INTEGER(v);
-    for (int i=0; i<n; i++) vd[i] = NA_INTEGER;
+    for (int i=from; i<=to; ++i) vd[i] = NA_INTEGER;
   } break;
   case REALSXP : {
-    double *vd = REAL(v);
-    for (int i=0; i<n; i++) vd[i] = NA_REAL;
+    if (INHERITS(getAttrib(v, R_ClassSymbol), char_integer64)) {
+      int64_t *vd = (int64_t *)REAL(v);
+      for (int i=from; i<=to; ++i) vd[i] = INT64_MIN;
+    } else {
+      double *vd = REAL(v);
+      for (int i=from; i<=to; ++i) vd[i] = NA_REAL;
+    }
   } break;
   case STRSXP :
     // character columns are initialized with blank string (""). So replace the all-"" with all-NA_character_
     // Since "" and NA_character_ are global constants in R, it should be ok to not use SET_STRING_ELT here. But use it anyway for safety (revisit if proved slow)
-    for (int i=0; i<n; i++) SET_STRING_ELT(v, i, NA_STRING);
+    // If there's ever a way added to R API to pass NA_STRING to allocVector() to tell it to initialize with NA not "", would be great
+    for (int i=from; i<=to; ++i) SET_STRING_ELT(v, i, NA_STRING);
     break;
   case VECSXP :
     // list columns already have each item initialized to NULL
     break;
   default :
-    error("Unsupported type '%s'", type2char(type));
+    error("Internal error: writeNA pass a vector of type '%s'", type2char(TYPEOF(v)));
   }
+}
+
+
+SEXP allocNAVector(SEXPTYPE type, R_len_t n)
```

Comment:
unless it complicates a lot, maybe we could switch to `R_xlen_t` already?

Reply:
A bit too much to juggle right now. It probably won't be more complicated, but there's a chance it will, and that unkown adds slightly to the cognitive burden. Thinking we'll just do a global replace in the next release and have to sweep through everything (including all `int`) then.